Menus/UndoMenu/UndoMenu.cs

```diff
@@ -0,0 +1,204 @@
+{
+    sealed class UndoMenu : MonoBehaviour, IInstantiateUI, IAlternateMenu, IUsesMenuOrigins, ICustomActionMap,
+        IControlHaptics, IUsesNode, IConnectInterfaces, IRequestFeedback, IUsesDeviceType
+    {
+        const float k_UndoRedoThreshold = 0.5f;
+        const float k_EngageUndoAfterStickReleasedDuration = 0.1f; // Duration after releasing the joystick to still accept a left/right flick to undo/redo.
+        const string k_FeedbackHintForJoystickController = "Click + flick left/right to undo/redo";
+        const string k_FeedbackHintForTrackpadController = "Click left/right side to undo/redo";
+
+        [SerializeField]
+        ActionMap m_ActionMap;
+
+        [SerializeField]
+        UndoMenuUI m_UndoMenuPrefab;
+
+        [SerializeField]
+        HapticPulse m_UndoPulse;
+
+        UndoMenuUI m_UndoMenuUI;
+        List<ActionMenuData> m_MenuActions;
+        Transform m_AlternateMenuOrigin;
+        MenuHideFlags m_MenuHideFlags = MenuHideFlags.Hidden;
+        float m_PrevNavigateX;
+        bool m_StillEngagedAfterStickRelease;
+        Coroutine m_StillEngagedAfterStickReleasedCoroutine;
+        bool m_TrackpadController;
+        string m_FeedbackHintForCurrentController;
+
+        readonly BindingDictionary m_Controls = new BindingDictionary();
+        
+        public Transform rayOrigin { private get; set; }
+
+        public Transform menuOrigin { get; set; }
+
+        public GameObject menuContent { get { return m_UndoMenuUI.gameObject; } }
+
+        public Node node { get; set; }
+
+        public event Action<Transform> itemWasSelected;
+
+        public Bounds localBounds { get { return default(Bounds); } }
+
+        public ActionMap actionMap { get { return m_ActionMap; } }
+        public bool ignoreLocking { get { return false; } }
+
+        public List<ActionMenuData> menuActions
+        {
+            get { return m_MenuActions; }
+            set
+            {
+                m_MenuActions = value;
+
+                if (m_UndoMenuUI)
+                    m_UndoMenuUI.actions = value;
+            }
+        }
+
+        public Transform alternateMenuOrigin
+        {
+            get { return m_AlternateMenuOrigin; }
+            set
+            {
+                m_AlternateMenuOrigin = value;
+
+                if (m_UndoMenuUI != null)
+                    m_UndoMenuUI.alternateMenuOrigin = value;
+            }
+        }
+
+        public MenuHideFlags menuHideFlags
+        {
+            get { return m_MenuHideFlags; }
+            set
+            {
+                if (m_MenuHideFlags != value)
+                {
+                    m_MenuHideFlags = value;
+
+                    var visible = value == 0;
+                    if (m_UndoMenuUI)
+                        m_UndoMenuUI.visible = visible;
+
+                    if (visible)
+                        ShowFeedback();
+                    else
+                        this.ClearFeedbackRequests();
```

Comment:
So... this doesn't allocate, but it does show up when I'm deep profiling (iterates over all feedback). I don't think it's a major worry but we should avoid spamming this method every frame. Feel free to punt until after this PR.